pages/blog/internals-of-nova-part-2.md

```diff
@@ -0,0 +1,286 @@
+---
+title: Internals of Nova Part 2 - Rows for the Row God, Columns for the Column Throne!
+description: Looking at more of the secret sauce that makes Nova.
+date: 2024-09-14
+authors:
+  - name: Aapo Alasuutari
+    url: https://github.com/aapoalas
+---
+
+[Last time](./internals-of-nova-part-1.md) I talked about how non-ordinary objects in Nova delegate their object features to a "backing object". This allows "exotic objects" to focus on the features that they are meant for and not get bogged down in the details that is JavaScript objects. This saves us some memory on every exotic object, but there is nothing particularly amazing about this trick. Any old engine could do the same thing and reap the same benefits.
+
+This time I will delve into the foremost idea behind Nova's heap structure; this idea is also what has driven me to dedicate my time to Nova. The name of this idea is the "heap vector" but more generally what I am talking about is [entity component system](https://en.wikipedia.org/wiki/Entity_component_system) and [data-oriented design](https://en.wikipedia.org/wiki/Data-oriented_design).
+
+## Data-oriented design
+
+Data-oriented design is a software design pattern or philosophy originating from the games industry, although one could perhaps argue that data-oriented design is the original design philosophy of software in general. In an abstract sense, a program is nothing more than input data and transformations that act upon said input data. A program's function is then to perform those transformations and do it as efficiently as possible (with some constraints, eg. developer time) to produce correct output from the input data.
+
+When we write a program, it is then fairly obvious that we cannot design the correct transformations without knowing what our input and output is, and more importantly we cannot design the transformations to be efficient without knowing our input data. As the input changes, the transformations must change and thus our program must change.
+
+Data-oriented design's first tenet is thus that we must know our data. If we do not know our data, we do not know the transformations we need to produce correct output, and that transformation is eventually the program that we must write. If we do not know our data, we do not know our problem. As developers, our job is not to write software; our job is to write programs that solve the problem at hand. If we do not know our data, we cannot do our job.
```

Comment:
**praise**: This section sounds a bit like Mike Actons famous talk 😉.

**suggestion**: An idea would be to qoute or reference his talk here as a "hook" for this section?

Reply:
Yeah, I was definitely calling back to it in my own words. Then I started watching the talk again and found that I shouldn't try saying it in my own words so much and should just copy him more :D